pandas/io/sql.py

```diff
@@ -778,16 +778,29 @@ def _sqlalchemy_con(connectable, need_transaction: bool):
     sqlalchemy = import_optional_dependency("sqlalchemy", errors="raise")
 
     if isinstance(connectable, str):
-        connectable = sqlalchemy.create_engine(connectable)
-    if isinstance(connectable, sqlalchemy.engine.Engine):
+        engine = sqlalchemy.create_engine(connectable)
+        try:
+            with engine.connect() as con:
+                if need_transaction:
+                    with con.begin():
+                        yield con
+                else:
+                    yield con
+        finally:
+            engine.dispose()
```

Comment:
This will dispose a user provided `Engine`? If so, I think the public sql functions will need to update `con` to note that they will be disposed now

Reply:
Oh nevermind, this is a str connection (looked at the diff wrong)